source/fuzz/equivalence_relation.h

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SOURCE_FUZZ_EQUIVALENCE_RELATION_H_
+#define SOURCE_FUZZ_EQUIVALENCE_RELATION_H_
+
+#include <map>
+#include <memory>
+#include <set>
+#include <unordered_set>
+#include <vector>
+
+#include "source/util/make_unique.h"
+
+namespace spvtools {
+namespace fuzz {
+
+// A class for representing an equivalence relation on objects of type |T|,
+// which should be a value type.  The type |T| is required to have a copy
+// constructor, and the |PointerHashT| and |PointerEqualsT| must be functors
+// providing hashing and equality testing functionality for pointers to objects
+// of type |T|.
+template <typename T, typename PointerHashT, typename PointerEqualsT>
```

Comment:
Somewhere (maybe not here) it would be great to have a comment (or comments) describing a few more details, including the implementation a bit. It is a disjoint-set (a.k.a. union-find, merge-find) data structure. There are no weight/rank/size values. There is path compression. The disjoint sets are represented as trees, where the root is the representative of the set. [Maybe stop here?] Getting the representative element simply requires following the parent pointer of an element until you reach the root. Checking equivalence of two elements requires checking that the representatives are equal.

This paragraph is perhaps more for the user of the class: PointerHashT and PointerEqualsT are used to detect trivial equality between elements and otherwise are NOT used to define the equivalence relation (although elements that are trivially equal and are in the EquivalenceRelation WILL be deemed equivalent). Perhaps this is obvious. The equivalence relation is given by manually adding pairs of T that are deemed to be equivalent (and are most likely NOT trivially equal, although they can be). For example, 1 and 5 might be added as equivalent, so that IsEquivalent(1,5) holds, because they represent IDs in a SPIR-V binary that contain the same value at run time, but clearly 1 != 5. Yet, 1 and 1 are trivially equal, so IsEquivalent(1,1) will also hold.

Each unique (in terms of equality) value added is copied into `owned_values_`, so there is one canonical memory address per unique value. Uniqueness is ensured by storing (and checking) a set of pointers to these values in `value_set_`, which uses `PointerHashT` and `PointerEqualsT`. `parent_` and `children_` encode the equivalence relation, i.e., the trees.

Reply:
Done, thanks!